buildreport/buildreport.go

```diff
@@ -0,0 +1,453 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// Package buildreport helps maintain tracking issues containing a list of build status entries in a
+// table that may be updated concurrently by multiple data sources. This is used in the release
+// process to keep track of builds for multiple releases happening at the same time.
+package buildreport
+
+import (
+	"context"
+	_ "embed"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/go-github/github"
+	"github.com/microsoft/go-infra/gitcmd"
+	"github.com/microsoft/go-infra/githubutil"
+	"github.com/microsoft/go-infra/stringutil"
+)
+
+// Symbols are used to indicate report entry status.
+const (
+	SymbolFailed     = "❌"
+	SymbolSucceeded  = "✅"
+	SymbolInProgress = "🏃"
+	SymbolNotStarted = "⌚"
+)
+
+const symbolKey = "" + SymbolNotStarted + " Waiting for first report, " +
+	SymbolInProgress + " In progress, " +
+	SymbolFailed + " Failed, " +
+	SymbolSucceeded + " Succeeded"
+
+const dataSectionMarker = "section generated by go-infra './cmd/releasego report'."
+const beginDataSectionMarker = "<!-- BEGIN " + dataSectionMarker + " -->"
+const endDataSectionMarker = "<!-- END " + dataSectionMarker + " -->"
+
+const beginDataMarker = "<!-- DATA "
+const endDataMarker = " DATA -->"
+
+const reportRetryTimeoutDuration = time.Minute
+const reportRetryDelayDuration = 3 * time.Second
+
+const githubWikiDefaultBranch = "refs/heads/master"
+const localTempBranch = "refs/heads/buildreport-temp"
+
+const releaseBuildPipelineName = "microsoft-go-infra-release-build"
+const releaseImagesPipelineName = "microsoft-go-infra-release-go-images"
+
+var pipelinesWithRetryInstructions = map[string]struct{}{
+	releaseBuildPipelineName:  {},
+	releaseImagesPipelineName: {},
+}
+
+// Update updates the report then sends a notification comment if necessary.
+func Update(ctx context.Context, owner, repoName, pat string, issue int, s State) error {
+	if err := UpdateIssueBody(ctx, owner, repoName, pat, issue, s); err != nil {
+		return err
+	}
+	return Notify(ctx, owner, repoName, pat, issue, s)
+}
+
+// UpdateIssueBody updates the given issue with new state. Requires the target GitHub repo to have
+// the wiki activated to perform safer concurrent updates than a simple issue description edit.
```

Comment:
I did a simple retry at first, but no. 😄 I spawned some goroutines to run edits at the same time and it hit issues immediately.

The reason GitHub is able to detect an edit conflict in the web UI is because the issue description editor is a form, and it uses some kind of form validation token/data that somehow stores info about the "current" state of the issue as far as that client is aware. If the form's "current" data doesn't match the authoritative current data on GitHub, GitHub refuses to run the update.

The GitHub API doesn't send any "current" state, only the target body text. So GitHub has no way to tell if one API-based edit conflicts with another, and the last one wins.

Passing an ETag (["If-Match"](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Match)) or date ("If-Unmodified-Since") header are the normal ways to make this work in REST APIs, but neither works here. GitHub's ETags are weak and can't be used this way, and "If-Unmodified-Since" does nothing at all.

I wasn't able to find *any* GitHub API that has concurrency safety that we could reasonably use. The GitHub API that *should* behave like a safe Git push isn't even safe and silently performs the equivalent of a force push. I assume this is because it would be expensive/hard to make any concurrent-safe API work with a distributed system, so they didn't. 😅 

But once we go outside the GitHub API, `git push` works fine. I didn't want to spam everyone with new `ref`s on the main microsoft/go repo, and the wiki happens to also be a Git repo with permissions that are linked to microsoft/go permissions, so it seemed reasonable.